tests/unit/test_network_auth.py

```diff
@@ -320,7 +325,7 @@ def _send(sent_req: MockRequest, **kwargs: Any) -> MockResponse:
     resp.status_code = 401
     resp.connection = connection
 
-    auth.handle_401(resp)
+    auth.handle_401(cast("Response", resp))
```

Comment:
To avoid this cast, perhaps adapt https://github.com/pypa/pip/blob/cc1ad642092a4928fd71e894089bfe68bfe21325/tests/lib/requests_mocks.py to define MockRequest in a `TYPE_CHECKING` block:

```
if TYPE_CHECKING:
    from requests.models import Response as MockResponse
else:
    class MockResponse:
        request: MockRequest
        connection: MockConnection
        url: str
```

Reply:
I prefer the current implementation over "action at a distance" effects that make the type checker think something different is happening than what is actually happening.

Reply:
@pradyunsg Are you referring to @pelson comment or my change?

Currently mypy is checking the [wrong types](https://github.com/pypa/pip/pull/13476#discussion_r2216179847), so something needs to be fixed.

I've not yet looked at @pelson's specific suggestion here.